xdsl/builder.py

```diff
@@ -99,14 +107,160 @@ def region(input: Callable[[Builder], None]) -> Region:
 
     @staticmethod
     def region(
-        input: list[Attribute] | ArrayAttr[Attribute]
+        input: Sequence[Attribute] | ArrayAttr[Attribute]
         | Callable[[Builder], None]
     ) -> Callable[[_CallableRegionFuncType], Region] | Region:
         if isinstance(input, Callable):
             return Builder._region_no_args(input)
         else:
             return Builder._region_args(input)
 
+    @staticmethod
+    def _implicit_region_no_args(func: Callable[[], None]) -> Region:
+        """
+        Generates a single-block region.
+        """
+        block = Block()
+        builder = Builder(block)
+
+        with _ImplicitBuilder(builder):
+            func()
+
+        return Region.from_block_list([block])
+
+    @staticmethod
+    def _implicit_region_args(
+        input_types: Sequence[Attribute] | ArrayAttr[Attribute]
+    ) -> Callable[[_CallableImplicitRegionFuncType], Region]:
+        """
+        Decorator for constructing a single-block region, containing the implementation of a
+        region with some input arguments.
+        """
+
+        if isinstance(input_types, ArrayAttr):
+            input_types = input_types.data
+
+        def wrapper(func: _CallableImplicitRegionFuncType) -> Region:
+            block = Block.from_arg_types(input_types)
+            builder = Builder(block)
+
+            with _ImplicitBuilder(builder):
+                func(block.args)
+
+            region = Region.from_block_list([block])
+            return region
+
+        return wrapper
+
+    @overload
+    @staticmethod
+    def implicit_region(
+        input: Sequence[Attribute] | ArrayAttr[Attribute]
+    ) -> Callable[[_CallableImplicitRegionFuncType], Region]:
+        """
+        Annotation used to construct a Region tuple from a function.
+        The annotation can be used in two ways:
+
+        For regions that have inputs or outputs:
+        ```
+        @Builder.implicit_region(input_types)
+        def func(args: tuple[BlockArgument, ...]) -> None:
+            ...
+        ```
+
+        For regions that don't have inputs or outputs:
+        ``` python
+        @Builder.implicit_region
+        def func() -> None:
+            ...
+        ```
+        """
+        ...
+
+    @overload
+    @staticmethod
+    def implicit_region(input: Callable[[], None]) -> Region:
+        ...
+
+    @staticmethod
+    def implicit_region(
+        input: Sequence[Attribute] | ArrayAttr[Attribute]
+        | Callable[[], None]
+    ) -> Callable[[_CallableImplicitRegionFuncType], Region] | Region:
+        if isinstance(input, Callable):
+            return Builder._implicit_region_no_args(input)
+        else:
+            return Builder._implicit_region_args(input)
+
+
+# Implicit builders
+
+
+@dataclass
+class _ImplicitBuilderStack(threading.local):
```

Comment:
I am a bit surprised by the usage of threading here. Why do you need it? Doesn't Python have a GIL? So threading really only helps for multiple I/O bound tasks, which is not what we have here, right?

I guess my question boils down to why/where do you assume that we are running multi-threaded?

Reply:
The GIL helps make sure that there's no weirdness at a more granular level, so every instruction is correct. But if we're building two implicit regions in parallel, on different threads, the operation inits can be interleaved, and each should be added to the builder for its own thread. The stacks should be different, basically. As far as I can tell, the tests run in parallel, and that's the biggest opportunity for the shared mutable state to be an issue, or at least until we start doing some parallel compilation or whatever.

Reply:
Ah I see, testing runs in parallel, that was he thing I was missing :+1: